File: src/test/java/com/ternbusty/takoyaki/command/StartCommandTest.java
package com.ternbusty.takoyaki.command;

import com.ternbusty.takoyaki.ipc.NotifySocket;
import com.ternbusty.takoyaki.state.ContainerStatus;
import com.ternbusty.takoyaki.state.State;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;

import java.io.IOException;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

class StartCommandTest {

    private static final String ROOT = "/run/takoyaki";

    private State createdState() {
        State s = new State();
        s.id = "ctr-a";
        s.status = ContainerStatus.CREATED.value;
        s.pid = 4242;
        s.bundle = "/tmp/bundle";
        return s;
    }

    @Test
    void startingCreatedContainerSendsNotifyAndPersistsRunning() throws IOException {
        State st = spy(createdState());
        doReturn(st).when(st).refreshStatus();
        // withStatus(RUNNING) must return a State whose save() is harmless in
        // tests — spy it.
        State updated = spy(createdState());
        updated.status = ContainerStatus.RUNNING.value;
        doNothing().when(updated).save(anyString());
        doReturn(updated).when(st).withStatus(ContainerStatus.RUNNING);

        // Build a Spec with non-empty process.args. StartCommand validates
        // that process.args isn't empty before signalling — without this,
        // start refuses with "spec.process.args is missing or empty".
        com.ternbusty.takoyaki.spec.Spec spec = new com.ternbusty.takoyaki.spec.Spec();
        spec.process = new com.ternbusty.takoyaki.spec.Spec.Process();
        spec.process.args = java.util.List.of("/bin/true");

        try (MockedStatic<State> sm = mockStatic(State.class);
             MockedStatic<NotifySocket> nm = mockStatic(NotifySocket.class);
             MockedStatic<com.ternbusty.takoyaki.util.Json> jm =
                     mockStatic(com.ternbusty.takoyaki.util.Json.class)) {
            sm.when(() -> State.load(anyString(), anyString())).thenReturn(st);
            jm.when(() -> com.ternbusty.takoyaki.util.Json.readFile(any(), any()))
                    .thenReturn(spec);
            // pathFor is a pure path builder — let the real one run so the
            // sendStart verification below checks the canonical path.
            nm.when(() -> NotifySocket.pathFor(anyString())).thenCallRealMethod();

            int rc = StartCommand.run(ROOT, "ctr-a");

            assertEquals(0, rc);
            // NotifySocket.sendStart is the actual "go" signal across the
            // unix domain socket — verify it fired with the canonical path.
            nm.verify(() -> NotifySocket.sendStart(eq("/tmp/takoyaki-ctr-a.sock")));
            verify(updated, times(1)).save(eq("/run/takoyaki"));
        }
    }


    @Test
    void startingNonCreatedContainerErrors() {
        State st = spy(createdState());
        st.status = ContainerStatus.RUNNING.value;
        doReturn(st).when(st).refreshStatus();

        try (MockedStatic<State> sm = mockStatic(State.class);
             MockedStatic<NotifySocket> nm = mockStatic(NotifySocket.class)) {
            sm.when(() -> State.load(anyString(), anyString())).thenReturn(st);

            int rc = StartCommand.run(ROOT, "ctr-a");

            assertEquals(1, rc, "OCI: start only valid on a 'created' container");
            // No "go" signal should have been sent.
            nm.verifyNoInteractions();
        }
    }

    @Test
    void stateLoadFailureReturnsError() {
        try (MockedStatic<State> sm = mockStatic(State.class)) {
            sm.when(() -> State.load(anyString(), anyString()))
                    .thenThrow(new IOException("corrupt"));

            int rc = StartCommand.run(ROOT, "ctr-a");

            assertEquals(1, rc);
        }
    }
}

File: src/test/java/com/ternbusty/takoyaki/command/StartCommandTest.kt
package com.ternbusty.takoyaki.command

import com.ternbusty.takoyaki.ipc.NotifySocket
import com.ternbusty.takoyaki.state.ContainerStatus
import com.ternbusty.takoyaki.state.State
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockkObject
import io.mockk.spyk
import io.mockk.unmockkObject
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.io.IOException

class StartCommandTest {

    companion object {
        private const val ROOT = "/run/takoyaki"
    }

    private fun createdState(): State = State(
        id = "ctr-a",
        status = ContainerStatus.CREATED.value,
        pid = 4242,
        bundle = "/tmp/bundle",
    )

    @Test
    @Throws(IOException::class)
    fun startingCreatedContainerSendsNotifyAndPersistsRunning() {
        val st = spyk(createdState())
        every { st.refreshStatus() } returns st
        val updated = spyk(createdState().copy(status = ContainerStatus.RUNNING.value))
        every { updated.save(any()) } just Runs
        every { st.withStatus(ContainerStatus.RUNNING) } returns updated

        mockkObject(State.Companion, NotifySocket)
        try {
            every { State.load(any(), any()) } returns st
            every { NotifySocket.pathFor(any()) } answers { callOriginal() }
            every { NotifySocket.sendStart(any()) } just Runs

            val rc = StartCommand.run(ROOT, "ctr-a")

            assertEquals(0, rc)
            io.mockk.verify { NotifySocket.sendStart("/tmp/takoyaki-ctr-a.sock") }
            io.mockk.verify(exactly = 1) { updated.save("/run/takoyaki") }
        } finally {
            unmockkObject(State.Companion, NotifySocket)
        }
    }

    @Test
    fun startingNonCreatedContainerErrors() {
        val st = spyk(createdState().copy(status = ContainerStatus.RUNNING.value))
        every { st.refreshStatus() } returns st

        mockkObject(State.Companion, NotifySocket)
        try {
            every { State.load(any(), any()) } returns st

            val rc = StartCommand.run(ROOT, "ctr-a")

            assertEquals(1, rc, "OCI: start only valid on a 'created' container")
            io.mockk.verify(exactly = 0) { NotifySocket.sendStart(any()) }
        } finally {
            unmockkObject(State.Companion, NotifySocket)
        }
    }

    @Test
    fun stateLoadFailureReturnsError() {
        mockkObject(State.Companion)
        try {
            every { State.load(any(), any()) } throws IOException("corrupt")

            val rc = StartCommand.run(ROOT, "ctr-a")

            assertEquals(1, rc)
        } finally {
            unmockkObject(State.Companion)
        }
    }
}
